### DISCUSSION TITLE
Adding support to extension boards[External or daughter board for extra IO]

### DISCUSSION THREAD

**Discussion by muhammadelmogy** [grblHAL/core#291] [Ideas]
I was wondering if it is possible to add support for a daughter board for extra IO.
I know almost most commercial controllers boards always have other daughter boards that can be connected to the main board which can add extra IO to the main board.

This could be implemented by using another microcontroller or maybe one or more of those I2C  IO expander ICs like PCF8575
One can use limited IO IC like for example the stm32f411 which have floating point support and other good things and add IO expander boards to have extra useful IO without the need to design a complex board with a lot of IO  as a main board.

**Comment by terjeio** [comment]
Like [this](https://github.com/grblHAL/RP2040/blob/master/PCA9654E.c) or [this](https://github.com/grblHAL/ESP32/blob/master/main/ioexpand.c), used [here](https://github.com/grblHAL/ESP32/blob/f3b73ab4cb7f640724e29f606704f02dbce3d25c/main/driver.c#L1361-L1415) and other places in [driver.c.](https://github.com/grblHAL/ESP32/blob/master/main/driver.c).

Driver code can implement I/O as it sees fit - the core does not care.

**Comment by muhammadelmogy** [comment]
So I need to add an IC driver specific code (for example PCF8575) and an ioexpand.c to determine the expansion IO pins and also modify the driver code to adopt the new ADD IO in ioexpand.c ?
Am I Correct?

**Reply by terjeio** [reply]
> Am I Correct?

Yes, you may also add the code as board specific - depending on what you are going to use the pins for. [An example](https://github.com/grblHAL/STM32F4xx/blob/master/Src/st_morpo_dac.c), but not for i/O expansion...
